cert/cert_test.go

```diff
@@ -0,0 +1,132 @@
+package cert
+
+import (
+	"io/ioutil"
+	"strings"
+	"testing"
+
+	"github.com/cloudflare/certmgr/cert"
+)
+
+var (
+	ca1 = &cert.CA{
+		Name:        "test_ca",
+		Remote:      "test_remote",
+		Label:       "test_label",
+		Profile:     "test_profile",
+		AuthKey:     "1111",
+		AuthKeyFile: "test_keyfile",
+		File:        nil,
+	}
+	ca2 = &cert.CA{
+		Name:        "test_ca",
+		Remote:      "test_remote",
+		Label:       "test_label",
+		Profile:     "test_profile",
+		AuthKey:     "1111",
+		AuthKeyFile: "test_keyfile",
+		File:        nil,
+	}
+)
+
+func TestCompareCertificatesInnerNewlines(t *testing.T) {
```

Comment:
This is mostly personal preference, but `TestCompareCertificatesInnerNewlines` and `TestCompareCertificateOuterNewlines` are basically the same test but with different inputs - I'm a huge fan of "table driven tests" to combine similar tests such that they are more compact and readable - see https://dave.cheney.net/2013/06/09/writing-table-driven-tests-in-go for some good reading.
e.g. those 2 Test functions become:

```golang
func TestCompareCertificatesNewlinesEx1(t *testing.T) {
	var tests = []struct{
		name string
		path string
	}{
		{"inner newlines","test_files/inner-newlines.pem"},
		{"outer newlines","test_files/outer-newlines.pem"},
	}
	for _, tt := range tests {
		t.run(tt.name, func(t *testing.T){
			dat1, err := ioutil.ReadFile("test_files/google.pem")
			if err != nil {
				t.Fatal(err)
			}
			pem := string(dat1)

			dat2, err := ioutil.ReadFile(tt.path)
			if err != nil {
				t.Fatal(err)
			}
			pemNewline := string(dat2)

			ca1.SetPEM([]byte(pem))
			ca2.SetPEM([]byte(pemNewline))
			isSame, err := cert.CompareCertificates(ca1.GetPEM(), ca2.GetPEM())
			if err != nil {
				t.Fatal(err)
			}
			if !isSame {
				t.Fatal("Duplicate certificates do not match")
			}
		})
	}
}
```

The `require` package: https://godoc.org/github.com/stretchr/testify/require is very popular for eliminating repetitive assertions in tests, the addition of that slims down the test even further:


```golang
func TestCompareCertificatesNewlinesEx2(t *testing.T) {
	var tests = []struct{
		name string
		path string
	}{
		{"inner newlines","test_files/inner-newlines.pem"},
		{"outer newlines","test_files/outer-newlines.pem"},
	}
	for _, tt := range tests {
		t.run(tt.name, func(t *testing.T){
			dat1, err := ioutil.ReadFile("test_files/google.pem")
			require.NoError(t, err)
		
			pem := string(dat1)

			dat2, err := ioutil.ReadFile(tt.path)
			require.NoError(t, err)
			pemNewline := string(dat2)

			ca1.SetPEM([]byte(pem))
			ca2.SetPEM([]byte(pemNewline))
			isSame, err := cert.CompareCertificates(ca1.GetPEM(), ca2.GetPEM())
			require.NoError(t, err)
			require.True(t, isSame, "Duplicate certificates do not match")
		})
	}
}
```

Reducing from 50->25 lines might not be a lot now, but in the future adding a new testcases is as simple as adding a new item to the tests slice.

Same logic can be applied to the other tests :)

Reply:
Thanks for the tip! I'll update the similar tests